### ISSUE TITLE
Support for branching (if / then) operation

### ISSUE THREAD

**Issue by dylburger** [PipedreamHQ/pipedream#216]
Currently, workflows are a linear sequence of steps. I'd like support to branch a workflow in a conditional manner (if condition is true, run this step, else run this other step).

**Comment by rugbbyli** [comment]
Hello, are there any updates for this feature ? @dylburger

**Comment by dylburger** [comment]
@rugbbyli not yet but we’re still actively thinking about it

**Comment by ThinkFar** [comment]
@dylburger, I would love to see this. 
This is one of my biggest pain points with pipedream.

**Comment by kpetrillo** [comment]
Off the top of my head, this is my only request that would make my life so much easiest. Pipedream is a solid product but having conditional steps would be a game changer!

**Comment by essential-randomness** [comment]
Would it be possible to have some sort of "goto step #X if condition is true" block? I don't know how complex that would be to implement, but it seems in line with what other steps can do.

**Comment by dylburger** [comment]
@essential-randomness That's great feedback and yes, this would be possible with the implementation we envision for this.

**Comment by giondo** [comment]
no update yet? 
I would be great to have conditionals steps
so we can perform different actions with diff integrations depending on the trigger

**Comment by dylburger** [comment]
@giondo No update on this yet, but it's near the top of our backlog after we finish #220 (currently our #1 most-requested feature).

**Comment by EnGlamdring** [comment]
@dylburger +1 I'm doing multiple flows right now to handle this and it's driving me nuts. Thanks!

**Comment by devang-kredx** [comment]
I'd like to keep logic of different branches of a workflow separated and clean so that it's easier to track and debug (rather than managing branching in the code itself and then having if-else statement in every subsequent block)

**Comment by jonparker** [comment]
We would love this feature.

Even the ability to have a simple condition to skip a step.

We have some workflows that have 3 or 4 Slack steps but we aren't able to support proper testing which would require having the workflow use a test Slack account.

Currently the only way to do it is to either have two copies of the one workflow or duplicate and convert all the Slack steps into code steps and then have an if statement around each code block to only run it if the test flag is true for the test steps and false for the live steps.

I guess an alternative would be some kind of condition which determines which account to use for the `auths` object.

This would mean allowing multiple Slack accounts to be connected to a single step.

**Comment by LucBerge** [comment]
> I guess an alternative would be some kind of condition which determines which account to use for the auths object. This would mean allowing multiple Slack accounts to be connected to a single step.

@jonparker Great idea but from my point of view, **this is indenpendent from this issue**. You want the apps account props to be seen as pointers to an account. You could then change the pointer in the previous steps so that Slack account A or B is used to send a message.

When this issue will be closed, you can do a simple if condition with identical slack actions on each branch but different accounts.

For now, I recommend you to open a new issue.

**Comment by jonparker** [comment]
Hi @LucBerge , I am no longer using Pipedream in that team. I am only using it for small personal projects.

**Comment by yangcheng** [comment]
is there any update on this feature?

**Comment by modernmusician** [comment]
very important feature!

the "Router" module in Integromat / Make is a great reference.

one way you could improve on it is by adding sequencing to the router steps so it can either complete the branches sequentially, or you could do it synchronously

**Comment by b2storefront** [comment]
Very much waiting for this feature to be implemented, upvoting! :)

**Comment by elibroftw** [comment]
Kinda of thinking of replacing pipedream now with my website because of this. I keep getting errors because the source was accessed via a GET request and not  POST so my parsing step returns nothing to my Discord webhook step.

**Comment by yangcheng** [comment]
agree, . If I have to write code instead of releasing the power to non-engineers,  I'd go with lambda or cloudflare workers, way cheaper.

**Comment by s1248** [comment]
I think this is an extremely important feature. And now this is also my pain point when using Pipedream. Currently I'm working on using Microsoft Power Automate. To set conditions and switch each case will send requests to each flow separately in Pipedream. However, the trade-off is a 1s delay for each trigger. It would be great if Pipedream adds this feature soon to compete with Microsoft Power Automate.

![image](https://user-images.githubusercontent.com/76787070/212870603-e7726c89-b1a9-4f65-ad20-c399b23c6922.png)

**Comment by bacherjo** [comment]
Hi Pipedream, why is branching still not there? This thread dates back to mid 2021, when you were "actively thinking about it". We need a bit more speed in feature delivery.
For the other feature I am waiting for (loop over an array) [https://github.com/PipedreamHQ/pipedream/issues/193#issuecomment-1303208624](url) it is now three years with no progress.

**Comment by chahalamol** [comment]
also wondering on the status of this feature

**Comment by fev4** [comment]
Is the alternative to use other workflows? Like, the work starts on one workflow and depending on what happens sends the information to another workflow, and so it starts to branch out. I'm gonna give that a try, but it seems difficult to visualize.

**Comment by acreconsulting** [comment]
+1 following

**Comment by tonyxiao** [comment]
It would also be great if the branches can be executed in parallel as well.

**Comment by wwc-development** [comment]
Is this an upcoming feature? This would be really helpful

**Comment by yangcheng** [comment]
has been more than 3 years and no update, I am moving all my pipedream workflow to cloudflare workers.

**Comment by dylburger** [comment]
Hi @yangcheng , this is still on our roadmap. Does Cloudflare workers support branching logic in their function UI?

**Comment by Zewb** [comment]
I saw it mentioned in Slack, but looks like this is still in-progress with no ETA. Looks like it would be a great addition when it happens though!

**Comment by drakeballew** [comment]
Agreed! Meanwhile, I’ve found that you can do this and a lot more by handling the branching yourself in a custom code block if you break up your larger workflow into sub component workflows and then use HTTP requests to trigger the sub flows. I can share an example of what I’m doing if it would be helpful, lmk.

**Comment by VladymyrPylypchatin** [comment]
+1 for this feature. The only reason for considering make/integromat or a custom solution

**Comment by Zewb** [comment]
> 

Yeah, that would definitely help! Please share anyone you've got. :)

**Comment by VladymyrPylypchatin** [comment]
In the meantime, here're a few handy workarounds that helped me avoid the need for if/else. Maybe it would be helpful for someone.

1. Custom actions "find or add."
<img width="1002" alt="Screenshot 2023-06-11 at 08 05 46" src="https://github.com/PipedreamHQ/pipedream/assets/47158131/c032196c-41ce-4165-927a-4a431b3743fa">


2. Mapping of the actions
<img width="1778" alt="Screenshot 2023-06-11 at 08 07 47" src="https://github.com/PipedreamHQ/pipedream/assets/47158131/0f3e8114-61ec-4267-918e-29ef8f28bc0b">

**Comment by drakeballew** [comment]
Here's an [example of chaining together workflows using HTTP responses as triggers](https://pipedream.com/new?h=tch_m5oflo). You can use custom code blocks to do the logic branching. I removed some steps with the special sauce so unfortunately this code won't run out of the box, but it should give you an idea of how you can branch things out logically. Let me know if you have questions!

------- Original Message -------

**Comment by bacherjo** [comment]
all the posted suggestiosn for branching via code steps are valid and OK, but I think what is missing big time is "conditional step execution". I would most love to have this as generic option in any http request step. It could be "true" by default, so nothing changes for existing flows.
And then just have a condition to define (using properties of previous steps) and based on truo or false execute the step or skip it.

My practrical use case: I receive an order and then have to check if the customer record already exists.
If not, create customer, then create order.
If yes, just create the order for existing customer.
This would be so simple with a "conditional execution" option.

Since we do not have it, I had to build a "router" workflow, which searches for the customer and then conditionally branches (via codestep) to two other workflows, one with customer and order creation, one with just order creation.
So for this simple requirement i end up having three workflows, and a tedious coding step.

**Comment by HardIsEasy** [comment]
+1 I just came from Make.com where router is very easy to set up makes pipedream look way to complicated

**Comment by mroy-seedbox** [comment]
Another type of branching to consider: run two (or more) branches in parallel (fan out), and then with the ability to fan back in later.

Something that could support a workflow like this:
```
   ----- A1 ------- A2 ------- A3 ------- A4
          |          |          |
          B1 ------- B2         |
          |                     |
          C1 ------- C2 ------- C3
                     |
                     D1
```

**Comment by expozeur** [comment]
How is it possible that something so simple, like the ability for custom logic to skip a certain step based on a condition, is not built in to an integrations tool that is intending to compete with the big guys like Zapier and Make? I'm a new Pipedream evangelist, but this is just disappointing. And then to see that this is an ongoing conversation and that there have been plans to "fix" this for 4 years is just embarrassing.

**Comment by dylburger** [comment]
@expozeur we’re working on this now. This is a complex feature to get right. Try to be patient as we make it happen.

**Comment by nickidw** [comment]
@dylburger any update?

**Comment by JBratches** [comment]
@dylburger – would love an update on this if you have it. this would be essential to onboard our team onto pipedream.

**Comment by mahomedh** [comment]
I think getting this feature is just a [pipe dream](https://dictionary.cambridge.org/us/dictionary/english/pipe-dream)

**Comment by andrewjschuang** [comment]
Is it a dream coming true?

<img width="660" alt="image" src="https://github.com/PipedreamHQ/pipedream/assets/15385076/328d493d-3355-4dea-91ed-65f23c307743">

**Comment by jingconan** [comment]
Please pls adding this feature. I have been wanting this feature since I cofounded my last last startup. 2 yrs have passed, I am now building a new one and now blocked by the lack of same feature again.

**Comment by bacherjo** [comment]
this thread is in its 5th year. 5 years of waiting and promises. could the pipedream team at least share why this is not implemented yet?

**Comment by andrewjschuang** [comment]
We are actively working on this feature as of right now, appreciate everyone's patience and am as excited as you all are!

**Comment by jingconan** [comment]
Hi Andrew, thanks a lot for the update and this is wonderful! We are very looking forward to try it once it is out

**Comment by dannyroosevelt** [comment]
@psavkar shared an update in our Slack community today to let folks know that **we're very close to shipping the preview of our first branching operator (If/Else)**. Here's a short demo -- let us know what you think! 
https://youtu.be/vsgdBV3LG4E?si=X1GDOweqGNCQL91d

**Comment by nickidw** [comment]
That really looks great!

**Comment by andrewjschuang** [comment]
We shipped the preview -- check out [the announcement](https://pipedream-users.slack.com/archives/C01DFPZNKJN/p1722462744334219) if you missed it!

**Comment by Maynardnaze** [comment]
Works great in testing but kind of sad it's not included with the Basic Plan.

**Comment by bacherjo** [comment]
Yes a dream came true, but then I realized that my existing flow suddenly costs me triple the credits, it went up from 2 to 6 credits per invocation, just because I added one if/else branch. I run this flow several thousand times per month, so the credits now melt away. This is sad, as I am considering reverting to a coding step instead of using the nice visual editor for if/else. 
Is this an inteded?
regards, Johannes

**Comment by nickidw** [comment]
> Yes a dream came true, but then I realized that my existing flow suddenly costs me triple the credits, it went up from 2 to 6 credits per invocation, just because I added one if/else branch. I run this flow several thousand times per month, so the credits now melt away. This is sad, as I am considering reverting to a coding step instead of using the nice visual editor for if/else. Is this an inteded? regards, Johannes

For this reason I removed all branching code as soon as I saw what it did to credit consumption.

**Comment by bacherjo** [comment]
well, if you can't code, you pay. period.
